File: leetcode/733.flood-fill.py
# @leet start
class Solution:
    def floodFill(
        self, image: list[list[int]], sr: int, sc: int, color: int
    ) -> list[list[int]]:
        original = image[sr][sc]

        if color == original:
            return image

        num_rows = len(image)
        num_cols = len(image[0])

        stack = [(sr, sc)]
        image[sr][sc] = color

        while stack:
            row, col = stack.pop()

            for delta_row, delta_col in [(-1, 0), (1, 0), (0, -1), (0, 1)]:
                if (
                    0 <= row + delta_row < num_rows
                    and 0 <= col + delta_col < num_cols
                    and image[row + delta_row][col + delta_col] == original
                ):
                    image[row + delta_row][col + delta_col] = color

                    stack.append((row + delta_row, col + delta_col))

        return image


# @leet end

File: leetcode/733.flood-fill.ts
// @leet start
function floodFill(
  image: number[][],
  sr: number,
  sc: number,
  color: number,
): number[][] {
  const origColor = image[sr][sc];

  if (origColor === color) {
    return image;
  }

  const stack: [number, number][] = [[sr, sc]];

  while (stack.length) {
    const [r, c] = stack.pop()!;

    if (
      r >= 0 &&
      r < image.length &&
      c >= 0 &&
      c < image[0].length &&
      image[r][c] === origColor
    ) {
      image[r][c] = color;

      stack.push([r - 1, c], [r + 1, c], [r, c - 1], [r, c + 1]);
    }
  }

  return image;
}
// @leet end
